In [1]:
import numpy as np

# Process attributes: [Arrival Time, Estimated Treatment Time, Priority]
processes = [
    [0, 30, 3],
    [10, 20, 5],
    [15, 40, 2],
    [20, 15, 4]
]

# Time Quantum for Round Robin
time_quantum = 10

def calculate_waiting_turnaround_time(processes, completion_time):
    num_processes = len(processes)
    waiting_time = [0] * num_processes
    turnaround_time = [0] * num_processes

    for i in range(num_processes):
        turnaround_time[i] = completion_time[i] - processes[i][0]
        waiting_time[i] = turnaround_time[i] - processes[i][1]

    return waiting_time, turnaround_time

def fcfs_scheduling(processes):
    num_processes = len(processes)
    completion_time = [0] * num_processes
    current_time = 0

    for i in range(num_processes):
        current_time = max(current_time, processes[i][0])
        completion_time[i] = current_time + processes[i][1]
        current_time = completion_time[i]

    return completion_time

def sjf_scheduling(processes):
    num_processes = len(processes)
    completion_time = [0] * num_processes
    remaining_time = [processes[i][1] for i in range(num_processes)]
    current_time = 0

    while True:
        remaining_processes = [i for i in range(num_processes) if processes[i][0] <= current_time and remaining_time[i] > 0]
        if not remaining_processes:
            break

        shortest_process = min(remaining_processes, key=lambda x: remaining_time[x])
        current_time += 1
        remaining_time[shortest_process] -= 1

        if remaining_time[shortest_process] == 0:
            completion_time[shortest_process] = current_time

    return completion_time

def priority_scheduling(processes):
    num_processes = len(processes)
    completion_time = [0] * num_processes
    remaining_time = [processes[i][1] for i in range(num_processes)]
    current_time = 0

    while True:
        remaining_processes = [i for i in range(num_processes) if processes[i][0] <= current_time and remaining_time[i] > 0]
        if not remaining_processes:
            break

        highest_priority_process = min(remaining_processes, key=lambda x: processes[x][2])
        current_time += 1
        remaining_time[highest_priority_process] -= 1

        if remaining_time[highest_priority_process] == 0:
            completion_time[highest_priority_process] = current_time

    return completion_time

def round_robin_scheduling(processes, time_quantum):
    num_processes = len(processes)
    completion_time = [0] * num_processes
    remaining_time = [processes[i][1] for i in range(num_processes)]
    current_time = 0

    while True:
        all_processes_completed = True

        for i in range(num_processes):
            if processes[i][0] <= current_time and remaining_time[i] > 0:
                all_processes_completed = False
                if remaining_time[i] <= time_quantum:
                    current_time += remaining_time[i]
                    completion_time[i] = current_time
                    remaining_time[i] = 0
                else:
                    current_time += time_quantum
                    remaining_time[i] -= time_quantum

        if all_processes_completed:
            break

    return completion_time

# FCFS Scheduling
fcfs_completion_time = fcfs_scheduling(processes)
fcfs_waiting_time, fcfs_turnaround_time = calculate_waiting_turnaround_time(processes, fcfs_completion_time)

# SJF Scheduling
sjf_completion_time = sjf_scheduling(processes)
sjf_waiting_time, sjf_turnaround_time = calculate_waiting_turnaround_time(processes, sjf_completion_time)

# Priority Scheduling
priority_completion_time = priority_scheduling(processes)
priority_waiting_time, priority_turnaround_time = calculate_waiting_turnaround_time(processes, priority_completion_time)

# Round Robin Scheduling
rr_completion_time = round_robin_scheduling(processes, time_quantum)
rr_waiting_time, rr_turnaround_time = calculate_waiting_turnaround_time(processes, rr_completion_time)

# Calculate Average Waiting Time and Average Turnaround Time for each algorithm
avg_fcfs_waiting_time = np.mean(fcfs_waiting_time)
avg_fcfs_turnaround_time = np.mean(fcfs_turnaround_time)
avg_sjf_waiting_time = np.mean(sjf_waiting_time)
avg_sjf_turnaround_time = np.mean(sjf_turnaround_time)
avg_priority_waiting_time = np.mean(priority_waiting_time)
avg_priority_turnaround_time = np.mean(priority_turnaround_time)
avg_rr_waiting_time = np.mean(rr_waiting_time)
avg_rr_turnaround_time = np.mean(rr_turnaround_time)

# Print results
print("FCFS Completion Time:", fcfs_completion_time)
print("FCFS Waiting Time:", fcfs_waiting_time)
print("FCFS Turnaround Time:", fcfs_turnaround_time)
print("Average FCFS Waiting Time:", avg_fcfs_waiting_time)
print("Average FCFS Turnaround Time:", avg_fcfs_turnaround_time)

print("\nSJF Completion Time:", sjf_completion_time)
print("SJF Waiting Time:", sjf_waiting_time)
print("SJF Turnaround Time:", sjf_turnaround_time)
print("Average SJF Waiting Time:", avg_sjf_waiting_time)
print("Average SJF Turnaround Time:", avg_sjf_turnaround_time)

print("\nPriority Completion Time:", priority_completion_time)
print("Priority Waiting Time:", priority_waiting_time)
print("Priority Turnaround Time:", priority_turnaround_time)
print("Average Priority Waiting Time:", avg_priority_waiting_time)
print("Average Priority Turnaround Time:", avg_priority_turnaround_time)

print("\nRound Robin Completion Time:", rr_completion_time)
print("Round Robin Waiting Time:", rr_waiting_time)
print("Round Robin Turnaround Time:", rr_turnaround_time)
print("Average Round Robin Waiting Time:", avg_rr_waiting_time)
print("Average Round Robin Turnaround Time:", avg_rr_turnaround_time)


FCFS Completion Time: [30, 50, 90, 105]
FCFS Waiting Time: [0, 20, 35, 70]
FCFS Turnaround Time: [30, 40, 75, 85]
Average FCFS Waiting Time: 31.25
Average FCFS Turnaround Time: 57.5

SJF Completion Time: [30, 65, 105, 45]
SJF Waiting Time: [0, 35, 50, 10]
SJF Turnaround Time: [30, 55, 90, 25]
Average SJF Waiting Time: 23.75
Average SJF Turnaround Time: 50.0

Priority Completion Time: [70, 105, 55, 85]
Priority Waiting Time: [40, 75, 0, 50]
Priority Turnaround Time: [70, 95, 40, 65]
Average Priority Waiting Time: 41.25
Average Priority Turnaround Time: 67.5

Round Robin Completion Time: [85, 60, 105, 75]
Round Robin Waiting Time: [55, 30, 50, 40]
Round Robin Turnaround Time: [85, 50, 90, 55]
Average Round Robin Waiting Time: 43.75
Average Round Robin Turnaround Time: 70.0


->
FCFS (First-Come-First-Serve):

Average Waiting Time: Calculated as 31.23 minutes
Average Turnaround Time: Calculated as 57.5 minutes

SJF (Shortest Job First):

Average Waiting Time: Calculated as 23.0 minutes
Average Turnaround Time: Calculated as 50.0 minutes
.

Priority Scheduling:

Average Waiting Time: Calculated as 41.25 minutes
Average Turnaround Time: Calculated as 67.5 minutes


Round Robin (RR) with Time Quantum 10:

Average Waiting Time: Calculated as 43.75 minutes
Average Turnaround Time: Calculated as 70.0 minutes


In summary:
SJF (Shortest Job First) provides the most efficient scheduling with the lowest average waiting time and turnaround time. It prioritizes patients with shorter estimated treatment times, reducing overall waiting times.

FCFS (First-Come-First-Serve) has a relatively higher average waiting time compared to SJF, making it less efficient. It schedules patients based on their arrival times, which may lead to longer waiting times for patients with longer estimated treatment times.

Priority Scheduling takes into account the urgency level of patients, prioritizing those with higher urgency levels. While it ensures that urgent cases are attended to promptly, it has a higher average waiting time compared to SJF.

Round Robin (RR) with a time quantum of 10 minutes provides fair scheduling in terms of waiting time distribution. However, it has a higher average waiting time compared to SJF and may not effectively prioritize urgent cases.

